gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaSource.java

```diff
@@ -342,7 +370,12 @@ private long getPreviousOffsetForPartition(KafkaPartition partition, SourceState
         partition.getTopicName(), partition.getId()));
   }
 
-  private void getAllPreviousOffsets(SourceState state) {
+  // need to be synchronized as this.previousOffsets need to be initialized once
+  private synchronized void getAllPreviousOffsets(SourceState state) {
+    if(this.doneGettingAllPreviousOffsets.get()){
```

Comment:
Since this method is synchronized and is the only place that changes the value of `doneGettingAllPreviousOffsets`, you don't need to use AtomicBoolean. Just need to use a `volatile` boolean.